// @tanstack/solid-start-plugin/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const viteSolid = require("vite-plugin-solid");
const startPluginCore = require("@tanstack/start-plugin-core");
const schema = require("./schema.cjs");
function TanStackStartVitePlugin(opts) {
  const options = schema.getTanStackStartOptions(opts);
  return [
    startPluginCore.TanStackStartVitePluginCore(
      {
        framework: "solid",
        getVirtualServerRootHandler(ctx) {
          return `
import { toWebRequest, defineEventHandler } from '@tanstack/solid-start/server';
import serverEntry from '${ctx.serverEntryFilepath}';

export default defineEventHandler(function(event) {
  const request = toWebRequest(event);
  return serverEntry({ request });
});`;
        },
        getVirtualClientEntry(ctx) {
          return `
import { hydrate } from 'solid-js/web';
import { StartClient } from '@tanstack/solid-start';
import { createRouter } from '${ctx.routerFilepath}';

const router = createRouter();

hydrate(() => <StartClient router={router} />, document.body);`;
        },
        getVirtualServerEntry(ctx) {
          return `
import { createStartHandler, defaultStreamHandler } from '@tanstack/solid-start/server';
import { createRouter } from '${ctx.routerFilepath}';

export default createStartHandler({
  createRouter,
})(defaultStreamHandler);`;
        }
      },
      options
    ),
    viteSolid({ ...options.solid, ssr: true })
  ];
}
exports.TanStackStartVitePlugin = TanStackStartVitePlugin;
exports.tanstackStart = TanStackStartVitePlugin;
