// @tanstack/solid-router/dist/cjs/lazyRouteComponent.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Solid$1 = require("solid-js/web");
const Solid = require("solid-js");
const Match = require("./Match.cjs");
const ClientOnly = require("./ClientOnly.cjs");
function isModuleNotFoundError(error) {
  return typeof (error == null ? void 0 : error.message) === "string" && /Failed to fetch dynamically imported module/.test(error.message);
}
function lazyRouteComponent(importer, exportName, ssr) {
  let loadPromise;
  let comp;
  let error;
  const load = () => {
    if (typeof document === "undefined" && (ssr == null ? void 0 : ssr()) === false) {
      comp = () => null;
      return Promise.resolve(comp);
    }
    if (!loadPromise) {
      loadPromise = importer().then((res) => {
        loadPromise = void 0;
        comp = res[exportName ?? "default"];
        return comp;
      }).catch((err) => {
        error = err;
      });
    }
    return loadPromise;
  };
  const lazyComp = function Lazy(props) {
    if (error) {
      if (isModuleNotFoundError(error)) {
        if (error instanceof Error && typeof window !== "undefined" && typeof sessionStorage !== "undefined") {
          const storageKey = `tanstack_router_reload:${error.message}`;
          if (!sessionStorage.getItem(storageKey)) {
            sessionStorage.setItem(storageKey, "1");
            window.location.reload();
            return {
              default: () => null
            };
          }
        }
      }
      throw error;
    }
    if (!comp) {
      const [compResource] = Solid.createResource(load, {
        initialValue: comp,
        ssrLoadFrom: "initial"
      });
      return Solid$1.memo(compResource);
    }
    if ((ssr == null ? void 0 : ssr()) === false) {
      return Solid$1.createComponent(ClientOnly.ClientOnly, {
        get fallback() {
          return Solid$1.createComponent(Match.Outlet, {});
        },
        get children() {
          return Solid$1.createComponent(Solid$1.Dynamic, Solid$1.mergeProps({
            component: comp
          }, props));
        }
      });
    }
    return Solid$1.createComponent(Solid$1.Dynamic, Solid$1.mergeProps({
      component: comp
    }, props));
  };
  lazyComp.preload = load;
  return lazyComp;
}
exports.lazyRouteComponent = lazyRouteComponent;
